pinot-query-runtime/src/main/java/org/apache/pinot/query/mailbox/ReceivingMailbox.java

```diff
@@ -133,4 +152,15 @@ public void cancel() {
   public int getNumPendingBlocks() {
     return _blocks.size();
   }
+
+  private void notifyReader() {
+    Reader reader = _reader;
+    if (reader != null) {
```

Comment:
To be fair, there is a race condition here. Being paranoid, the following sequence is legal and problematic:
1. The mailbox is created by the sender side
2. A message is sent. But just after `_reader` is read...
3. The reader is registered
4. The reader polls the mailbox and found nothing
5. The reader blocks on the `notEmpty` condition
6. The mailbox executes this line, which is false because when read, _reader was null
7. The message is not sent
8. No more messages are sent by the sender
9. The reader is infinitely waiting for the condition

I'll try to fix that next week by adding a lock with an optimistic read, so performance should not be affected

Reply:
This race condition has been fixed by adding the following invariant:
- In order to call poll, `_reader` must be not null.

I've added a `Precondition` in poll that verifies that (as long as there is a single reader thread, which is something we assume).

Reply:
The actual reason is that 4 cannot happen because 2 already appended the message in the array